Claire & Claudias pizzabot

ClaireCaudwell opened the pull request:
Claire Caudwell and Claudia Kjerrgren's pizzabot. 

Link to my deployed project: 
https://mystifying-lovelace-84f38f.netlify.app/

My reflections on how this week's project turned out:
It was a hard project, but we felt that we made the pizza bot work to a certain level. 

Things I'd like to have clarified or explained in more detail:
We thought that it was nice that the project brief was laid out step by step in terms of the iterations, but it would be nice to be able to have clearer instructions when it comes to thinking about how to do the form. This is because in the beginning we thought that we should make a static form, but then after a while we realised that you can use the form elements to be then submitted which will then show the next question. Like how the prompts work but with html and JavaScript combined. The this would create an interactive robot that would ask you different questions. So an example of how to think about this and how this works would be great!

ClaireCaudwell committed:
added style tag into html.

ClaireCaudwell committed:
Updated the javascript file with the first 4 iterations and one conditional

ClaireCaudwell committed:
We updated the if else condition for iteration 3 which prompts the user to insert the name of the pizza they want to order etc.

ClaireCaudwell committed:
Added iteration 4 cooking time if condition

ClaireCaudwell committed:
Added functions for each iteration

ClaireCaudwell committed:
Updated the cooking time function

ClaireCaudwell committed:
Updated HTML & JavaScript as well as add css file

ClaireCaudwell committed:
Deleted the select and option sections of the form.

ClaireCaudwell committed:
Added 3 pizza images, Updated HTML with form elements and other html elements to be used when using innerhtml in javascript. In javascript edited some of the didifferent prompt sentances, deleted validateordername function call and the pizzaisonmenu variable that the function data was inserted into. Replaced the pizzaisonmenu if and else statements with a new one which allows for the onsubmit event to be actioned when run. This still gives a true or false if else statement but get's the value of the pizza the user has entered from what they input in the html input text box. Added a pizzaimage variable which will hopefully allow for us to switch between images shown depending upon the pizza selected. And a new function which is called when the user submits the information from the enteredpizzaname. In CSS updated the styling for the second form and also enabled some display: none css for some elements which will be shown when the javascript function is called.

ClaireCaudwell committed:
Added new pizza image. Updated the readme, Added a robot emoji which will act as the pizzabot robot. Added the pizza images in html. In javascript we added the pixture variables, where each one contained the image that is defined in the innerhtml. We also added a function and if else conditional which assigned an image to the specific pizza types if chosen by the user. Also added the plural  variable and if statement which helps in our sentances if the user orders one or more pizza.

ClaireCaudwell committed:
Added pizza-pattern jpeg for background image. Updated read me file

ClaireCaudwell committed:
Updated README

ClaireCaudwell committed:
Updated html

ClaireCaudwell committed:
Added netlify site to read me

puj commented on README.md:
```diff
@@ -1,13 +1,55 @@
+
+4. Once we got the functions to work together we started to edit the prompts and alerts so we could view the information in the DOM. We used the document.getElementById().innerhtml to connect both the javascript sections to the html. 
+
+5. For the end result we managed to get the form functions and onsubmit attribute of the form to trigger the next function and the same for the next function. This meant that when the user inputted information and submitted with the submit button it made the next form section appear. 
```
This worked out super well!  You should be proud! 👏

ClaireCaudwell replied:
Thank you! We did get some help with this :)

puj commented on code/index.html:
```diff
@@ -4,9 +4,72 @@
+          <!-- Second form where the user will input and submit how many pizzas they want to order -->
+          <form action="http://httpbin.org/anything" autocomplete="off" onsubmit="enteredPizzaAmount(event)" onclick="enteredPizzaAmount()" method="POST" class="form-2" id="formTwo">
+            <!-- Enter how many pizzas -->
+            <h3 id="howMany"></h3>
```
Naming suggestion: "quantityHeader"

puj commented on code/index.html:
```diff
@@ -4,9 +4,72 @@
+            <!-- Enter how many pizzas -->
+            <h3 id="howMany"></h3>
+            <label for="enterAmount">
+              <input type="text" id="enterAmount" name="enterAmount" placeholder="How many?" required></select>
```
Naming suggestion: "quantity"

puj replied:
Consider using `max` and `min` attributes.  Currently we can enter negative values

puj commented on code/index.html:
```diff
@@ -4,9 +4,72 @@
+        <!-- First form section where user will input username & name of pizza -->
+        <div class="cards">
+          <h3> &#129302 Before we get started with your order please enter your username:</h3> 
+          <form action="#" autocomplete="off" onsubmit="enteredPizzaName(event)" method="POST" id="formOne" class="form-1">
```
You don't need the `action` or `method` attribute here at all since you are using the `onsubmit` to control what happens next.

Those attributes are only used when sending data to a remote server via the default form functionality.  

Later, you will also be doing this in Javascript, which means you probably won't use these attributes again for quite some time.

ClaireCaudwell replied:
Ok perfect. We wrote it in the beginning as we had only this one form, but then we added the second form section. But noted, that we don't need it as we have the last form that does that!

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+
+//--------------------------------------------------------------------------//
+
+//Sentances showing in html via the getElementById()
+// Greeting to the user from the pizzabot
+document.getElementById("greeting").innerHTML = `&#129302 Hi! Welcome to Claire & Claudia's pizza ordering service. I'm pizza bot and I'm here to help you with your order.`;
+
+//Showing the menu
+document.getElementById("menu").innerHTML = `On the menu we have: <li>${vegetarian}: ${pizzaPrice}kr</li> <li>${hawaiian}: ${pizzaPrice}kr</li> <li>${pepperoni}: ${pizzaPrice}kr</li>`;
+
+//Message to the user asking them to choose a pizza from the menu
+document.getElementById("choosePizza").innerHTML = "&#129302 Ok great. Let's get started with your order! Please write the name of the pizza you want to order:";
```
Clear, good structure/separation 👍

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+//--------------------------------------------------------------------------//
+
+//Function compares if the parameter chosenPizza is true or false. Will return true if user inputs correct pizza name, otherwise will show message which is called further down in enteredPizzaValue
+const validateOrderName = (chosenPizza) => {
+    if (chosenPizza.toLowerCase() === vegetarian.toLowerCase() || chosenPizza.toLowerCase() === hawaiian.toLowerCase() || chosenPizza.toLowerCase() === pepperoni.toLowerCase()){
+        return true;
+    }
+    else { 
+        return false;
+    }
+};
```
There are a number of ways to do this: 
```javascript
const validateOrderName = (chosenPizza) => {
    // To prevent doing `toLowerCase` three separate times
    const lowerCaseChoice = chosenPizza.toLowerCase();
  
    // Calculate the `boolean` value
    const isValidChoice = lowerCaseChoice === vegetarian.toLowerCase() || 
              lowerCaseChoice  === hawaiian.toLowerCase() || 
              lowerCaseChoice  === pepperoni.toLowerCase();
    return isValidChoice;
};
```
```Javascript
const validateOrderName = (chosenPizza) => { 
    // Single line return example showing that you can return the boolean directly
    return chosenPizza.toLowerCase() === vegetarian.toLowerCase() || 
               chosenPizza.toLowerCase() === hawaiian.toLowerCase() || 
               chosenPizza.toLowerCase() === pepperoni.toLowerCase();

};
```

puj replied:
There are so many ways to write a simple function, but consider these alternatives and formulate an idea of what looks best, what's easiest to understand, and how to think about these types of functions in the future 👍

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+
+//Function calculating the cost of the pizza based on the pizzaAmount and pizzaPrice. Returnsg the amount when function is called further down
+const calculateTotalCost = (orderQuantity, pizzaPrice) => {
+    const orderTotal = (orderQuantity * pizzaPrice);
+    return orderTotal;     
```
Clean

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+
+//Function calculating the cooking time based on amount of pizzas ordered
+const calculateCookingTime = (pizzaAmount) => {
+    if (pizzaAmount <= 2) {
+        return 10;
+    } else if (pizzaAmount >= 3 && pizzaAmount <= 5) {
+        return 15;
+    } else {
+        return 20;
+    }
```
Alternative logic:
```javascript
    if (pizzaAmount <= 2) {
        return 10;
    } else if (pizzaAmount >= 6) {
        return 20;
    } else {
        return 15;
    }
```

The code does the same thing, although this isn't necessarily better or worse, the trade-offs are:
- The return values aren't ascending when you read them (10, 20, 15)
- There is only one conditional in the `else if`

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+};
+
+//Function that assigns an image to the type of pizza ordered
+pizzaImage = (orderName) => { 
+    if (orderName.toLowerCase() === vegetarian.toLowerCase()) {
+        return vegPic;
+    } else if (orderName.toLowerCase() === pepperoni.toLowerCase()) {
+        return pepPic;
+    } else {
+        return hawPic;
+    }
+};
```
Nice, and good comment

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+//Function which is connected to the first form in html inside the onsubmit() attribute of the form. When the user inputs the username and the pizza they want and presses the submit button, if true (i.e. if pizza name matches what's on the menu), it runs the code and also shows the next prompt of how many pizzas they want. If that is not true then the they get the wrongPizza message asking them t ...
+const enteredPizzaName = (event) => {
+    event.preventDefault();//Stops the page from submitting information to the backend.
+    let orderName = document.getElementById("enterPizza").value;
+    let correctPizzaName = validateOrderName(orderName);
```
Both of these variables should be `const` since their values are not changing

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+    document.getElementById("formTwo").style.display = "flex";
+
+    //Defined variable and if statement which helps to change if the pizza if plural or not in the final order statement.
+    let plural = "'s";
+    if (orderAmount === "1") {
+        plural = "";
+    }
```
Yes yes yes yes. And more yes. 💯

ClaireCaudwell replied:
Thanks! We got a bit of help with this so can't really take all the credit :) It is a good thing to know tho!

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+const enteredPizzaAmount = (event) => {
+    event.preventDefault(); //Stops the page from submitting information to the backend.
+    let orderAmount = document.getElementById("enterAmount").value; //Taking the value (the amount) the user has inputted in the DOM for the amount of pizzas they want to order and put this inside a function to then use to calculate the cooking time below. 
+    let correctOrderAmount = calculateCookingTime(orderAmount);
+    let totalPrice = calculateTotalCost(orderAmount, pizzaPrice);
```
Very nice function naming `calculateCookTime` and `calculateTotalCost`

Taking this one step further, when you name your function `calculateXXXxxxx` your variable name can then easily match the function being called:
```javascript
    const cookingTime = calculateCookingTime(orderAmount);
    const totalCost = calculateTotalCost(orderAmount, pizzaPrice);
```

puj commented on code/pizzabot.js:
```diff
@@ -1,7 +1,140 @@
+
+//Changed the javascript code so it doesn't only prompt or print the information to the DOM. Now it prints certain questions based on user input. 
+//Functions being called when user inputs information and clicks submit. Before javascript was being run based on prompts which automatically waited for user input. But now javascript is being run on demand when an event happens(call back) and for this to work you need to seperate the parts of the programs into functions. And this is why the code has been moved around (refactored).
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
```
Seems to be a lot of whitespace here.  Maybe we can remove that?

puj commented on code/style.css:
```diff
@@ -0,0 +1,145 @@
+    margin: 20px 0px 10px 0;
+}
+
+.form-1 {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
+}
+
+.form-1 input {
+    text-align: center;
+    padding: 15px;
+    border: none;
+    font-size: 18px;
+}
```
`form-1`  is quite a generic name.  If I am reading the CSS and I'm trying to remember "Which one was `form-1` and which one was `form-2`, then a name like `form-quantity` could be helpful?

puj commented on code/index.html:
```diff
@@ -4,9 +4,72 @@
+            <label for="enterPizza">
+              <input type="text" id="enterPizza" name="enterPizza" placeholder="Type in pizza flavour" required></select>
+            </label>
+              <input type="submit" name="submit" value="Enter" class="submit-button">
```
Clicking this will trigger the event defined in the `onsubmit` attribute in the form.

puj reviewed (dismissed):
Hi Claire and Claudia!
Excellent job! I love the fact that the pizza bot reveals a new section at each stage of the process.  Well done.  Your code comments are of high quality, continue to pursue these high quality comments.  Good attention to detail.  One thing I noticed is that when your code became more complex, you became more structured.  Very very good.  In some places your variable/function naming conventions were very solid, try to let this propagate to the rest of your code :+1:



Things to fix: 
- let => const in many places
- Disallow negative values for pizza quantity

To update the PR:
- Make code changes 
- `git add -u`
- `git commit -m "Fixed xxxxxxx"`
- `git push`/`git push origin master`
-  Your PR will be up-to-date

ClaireCaudwell committed:
Updated JS, HTML & CSS file with requested changes
1. JS File: Changed some of the variables that were set to const to let. I changed them on the ones I believe one would want to reassign in the future. Also updated the js file with the id name changes suggested.
2. HTML file: Changed the input type from text to number where the user is meant to input how many pizzas they want. Also added the min and max values. Changed the form id and classes in order to be able to differentiate easier between them. Also updated the enterAmount id and names to quanity and quanityHeader as suggested.
3. Updated all name changes in HTML in CSS also.
4. Removed the whitespace.

ClaireCaudwell commented:
Thanks for the feedback. All changes made and PR sent.

puj commented:
@ClaireCaudwell I don't see your changes reflected here.  Am I missing something?  
You should be able to 

- Make code changes
- `git add -u`
- `git commit -m "Fixed xxxxxxx"`
- `git push/git push origin master`
- Your PR will be up-to-date

ClaireCaudwell commented:
> @ClaireCaudwell I don't see your changes reflected here. Am I missing something?
> You should be able to
> 
> * Make code changes
> * `git add -u`
> * `git commit -m "Fixed xxxxxxx"`
> * `git push/git push origin master`
> * Your PR will be up-to-date

No I don't think so. I made the changes to my code files and then used the git make code changes reference to re-push them to the master. I think this was on the 19th of Sept. I will re-push following the steps you've outlined. Should the changes show here?

ClaireCaudwell committed:
Added semi-colons to the variables for the different pizza types.

ClaireCaudwell commented:
Looks like I pushed the changes o the 16th of September, but they only pushed to my repo https://github.com/ClaireCaudwell/project-pizza-bot/commit/4bb7e1f3764c47d69e23016da87778f491c95c1c
 Just tried to do it again from the terminal in vsc. Used all the steps and for the last one git push origin master, as it didn't seem to like "git push/git push origin master". However nothing seems to have updated here.

Make code changes
git add -u
git commit -m "Fixed xxxxxxx"
git push/git push origin master

ClaireCaudwell committed:
Items fixed after code review:
1. JS File: Changed some of the variables that were set to const to let. I changed them on the ones I believe one would want to reassign in the future. Also updated the js file with the id name changes suggested.
2. HTML file: Changed the input type from text to number where the user is meant to input how many pizzas they want. Also added the min and max values. Changed the form id and classes in order to be able to differentiate easier between them. Also updated the enterAmount id and names to quanity and quanityHeader as suggested.
3. Updated all name changes in HTML in CSS also.
4. Removed the whitespace.

ClaireCaudwell committed:
Update README.md

ClaireCaudwell committed:
Update README.md
